tool/tsh/common/vnet_darwin.go

```diff
@@ -48,18 +49,41 @@ func (c *vnetCommand) run(cf *CLIConf) error {
 		return trace.Wrap(err)
 	}
 
-	ctx, cancel := context.WithCancelCause(cf.Context)
-	defer cancel(nil)
+	ctx, cancel := context.WithCancel(cf.Context)
+	defer cancel()
+	g, ctx := errgroup.WithContext(ctx)
 
-	manager, adminCommandErrCh, err := vnet.Setup(ctx, ctx, appProvider)
+	socket, socketPath, err := vnet.CreateSocket(ctx)
 	if err != nil {
-		if errors.Is(err, context.Canceled) {
-			return nil
-		}
 		return trace.Wrap(err)
 	}
 
-	err = vnet.Run(ctx, cancel, manager, adminCommandErrCh)
+	g.Go(func() error {
+		<-ctx.Done()
+
+		return trace.Wrap(socket.Close())
+	})
+
+	ipv6Prefix, err := vnet.IPv6Prefix()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	dnsIPv6 := vnet.Ipv6WithSuffix(ipv6Prefix, []byte{2})
+
+	g.Go(func() error {
+		return trace.Wrap(vnet.ExecAdminSubcommand(ctx, socketPath, ipv6Prefix.String(), dnsIPv6.String()))
+	})
+
+	manager, err := vnet.Setup(ctx, appProvider, socket, ipv6Prefix, dnsIPv6)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	g.Go(func() error {
+		return trace.Wrap(manager.Run(ctx))
+	})
+
+	err = g.Wait()
```

Comment:
to be honest the exported API is still a bit ugly to deal with in both tsh and connect

I think ideally for clients it would look something like this:

```go
// tsh
vnetManager, err := vnet.Setup(ctx, appProvider)
if err != nil {
  return trace.Wrap(err)
}
return trace.Wrap(vnetManager.Run(ctx))
```

```go
// connect
vnetManager, err := vnet.Setup(ctx, appProvider)
if err != nil {
  return trace.Wrap(err)
}
vnetRunCtx, vnetCancel := context.WithCancel(context.Background())
go func() {
  err := vnetManager.Run(vnetRunCtx)
  // TODO: report error or early termination somehow
}()
```

The admin subcommand would be started in `vnet.Setup` and get to the point where the password prompt completed succesfully and we got the TUN, that's everything you might want to cancel in the request context. From that point managing the lifecycle of the admin process would be handled within the `vnetManager`, which would also hide implementation details like the socket. I think a simplified exported API like that would be much nicer if it's doable.

the downside is it's stuffing more things into `vnet.Manager` (which i know was not a good name to start with). We could either do that, or rename the current `vnet.Manager` to something like `vnet.NetworkStack`, and make a new `vnet.ProcessManager` that manages the lifetime of the admin process as well as the `NetworkStack`. There would be some options there but I think that's less important than getting the exported API right

let me know if I'm making sense

Reply:
I like the idea of the introducing `NetworkStack` and `ProcessManager`. The proposed API does look nice, but I don't think it's possible to have it look exactly like this. `vnet.Setup` in Connect cannot take `ctx`, because that context is bound to the RPC lifetime and it would stop the admin process right after the RPC ends. The admin process needs to be executed with a long-running context.

But this gives me an idea, what if `vnetManager.Run` just didn't accept a context and we followed the API of something like `net/http.Server`?

We could have something like this:

```go
// connect
// Still accepts ctx – it's needed when waiting for the TUN device to be created.
// But the admin subcmd and other stuff is spawned using a background context
// that's not exposed outside of processManager.
success := false
processManager, err := vnet.Setup(ctx, appProvider)
if err != nil {
  return trace.Wrap(err)
}
defer func() {
  if !success {
    err := processManager.Close()
    // TODO: Log the error.
  }
}()
s.processManager = processManager
success = true

go func() {
  err := processManager.Run()
  // TODO: report error or early termination somehow
}()



// In the Stop RPC
s.processManager.Close()
```

I'll try to explore this idea, though idk yet if I'm going to branch off this PR or #41413. I like the API from this PR slightly more than the previous one, mostly because it clearly separates the long running stuff from other operations, without having to pass `adminCommandErrCh` around, etc.

Reply:
The only difference when comparing to something like `net/http.Server` would be that the caller is expected to call `processManager.Close()` not only after `processManager.Run`, but also after `vnet.Setup`. But I believe we can address this either through comments or function names.

Reply:
Alright, I cleaned it up, let me know what you think. `vnet.ProcessManager` ended up being an `errgroup.Group` on wheels, maybe we could even name it `TaskGroup` or something.

The biggest hiccup is around handling the cancellation of osascript prompt. The error we want to show to the user comes from `execAdminSubcommand`. We somehow have to make sure that the user sees that before the error about reading from a closed socket.

https://github.com/gravitational/teleport/blob/0cff6ea1f979ab5298c34b6449b7b19dc191ac13/lib/vnet/setup.go#L70-L106